### PR TITLE
Flat top windowing function

### PR DESCRIPTION
**Pull request by robbr48**
[Flat top windowing](https://en.wikipedia.org/wiki/Window_function#Flat_top_window) function. Implemented with the following parameters:
a0 = 1.0
a1 = 1.933
a2 = 1.286 
a3 = 0.388
a4 = 0.0322

### CODE REVIEW

File: HopsanGUI/Utilities/GUIUtilities.cpp [hunk 1]
```diff
@@ -395,6 +395,21 @@ void windowFunction(QVector<double> &data, WindowingFunctionEnumT function, doub
...
+        case FlatTopWindow: {
+            int N = data.size()-1;
+            rCa = 0;
+            for(int n=0; n<=N && N>0; ++n) {
+                double w = 1.0 - 1.933*cos(2*M_PI*n/N) + 1.286*cos(4*M_PI*n/N) - 0.388*cos(6*M_PI*n/N) + 0.0322*cos(8*M_PI*n/N);
+                data[n] *= w;
+                rCa += w;
+                rCb += w*w;
+            }
+            rCa /= N;
```
Comment (peterNordin) [inline HopsanGUI/Utilities/GUIUtilities.cpp:409]: Maybe need protection for division by zero in rCa rCb and rC?, the last one should it be b or c?

File: HopsanGenerator/include/generators/HopsanModelicaGenerator.h [hunk 2]
```diff
@@ -36,8 +36,8 @@ class HopsanModelicaGenerator : public HopsanGeneratorBase
...
+    void generateComponentObject(ComponentSpecification &comp, QString &typeName, QString &displayName, QString &cqsType, /*QStringList &initAlgorithms,*/ QStringList &preAlgorithms, QStringList &equations, QStringList &finalAlgorithms, QList<PortSpecification> &portList, QList<ParameterSpecification> &parametersList, QList<VariableSpecification> &variables, QTextStream &logStream, bool &success);
+    void generateComponentObjectNumericalIntegration(ComponentSpecification &comp, QString &typeName, QString &displayName, QString &cqsType, QStringList &initAlgorithms, QStringList &preAlgorithms, QStringList &equations, QStringList &finalAlgorithms, QList<PortSpecification> &portList, QList<ParameterSpecification> &parametersList, QList<VariableSpecification> &variables, QTextStream &logStream, bool &success);
```
Comment (peterNordin) [inline HopsanGenerator/include/generators/HopsanModelicaGenerator.h:40]: These bool &success are only added in the declaration but not in the definition. Leftovers ?

File: HopsanGenerator/src/generators/HopsanModelicaGenerator.cpp [hunk 3]
```diff
@@ -200,15 +200,15 @@ void HopsanModelicaGenerator::parseModelicaModel(QString code, QString &typeName
...
+                QString tempLine = line;
                 tempLine.remove(" ");
                 int idx = tempLine.indexOf("hopsanCqsType=");
                 cqsType = tempLine.at(idx+15);
             }
             else if(words.at(0) == "parameter")         //"parameter" keyword
             {
                 QString name = words.at(2).section("(",0,0);
-                QString unit = lines.at(l).section("unit=",1,1).section("\"",1,1);
+                QString unit = line.section("unit=",1,1).section("\"",1,1);
```
Comment (peterNordin) [inline HopsanGenerator/src/generators/HopsanModelicaGenerator.cpp:211]: I do not know modelica, but what about requirements on = does it have to be attached directly to the name (unit) or would it be valid modelica code with a space, like in 'unit ='

**Comment by peterNordin** [comment]
> [Flat top windowing](https://en.wikipedia.org/wiki/Window_function#Flat_top_window) function. Implemented with the following parameters:
> a0 = 1.0
> a1 = 1.933
> a2 = 1.286
> a3 = 0.388
> a4 = 0.0322

What make these parameters special? Are they always suitable ?

File: HopsanGUI/Utilities/GUIUtilities.cpp (hunk 1 above)
Comment (robbr48) [reply HopsanGUI/Utilities/GUIUtilities.cpp:421]: I added a check for empty data sets at the top of the function.

I believe the computations are correct. The formulas are:
<a href="https://www.codecogs.com/eqnedit.php?latex=C_a&space;=&space;\textup{mean}(\omega)&space;\\&space;C_b&space;=&space;\frac{\textup{mean}(\omega^2)}{\textup{mean}(\omega)^2}" target="_blank"></a>

File: HopsanGenerator/src/generators/HopsanModelicaGenerator.cpp (hunk 3 above)
Comment (robbr48) [reply HopsanGenerator/src/generators/HopsanModelicaGenerator.cpp]: Sorry, this was not supposed to be in this PR. I accidentally included a commit from another branch. But to answer the question, I don't know. All examples I have seen have the equality symbol attached to the name, but I guess spaces are allowed too. I will fix this in the other branch.

**Comment by robbr48** [comment]
The coefficients are from the ISO 18431-2 standard. There are other possible coefficients, Matlab for examples uses a different set. Ideally we could allow users to specify the coefficients, but that would require a lot of additional widgets and HCOM function arguments. It could even be put in the configuration XML file and accessed from the options dialog. But for now, I decided to go with he ISO standard.

File: HopsanGUI/Utilities/GUIUtilities.cpp [hunk 4]
```diff
@@ -381,6 +381,12 @@ QVector< complex<double> > realToComplex(const QVector<double> &rRealVector)
 //! @param [out] Cb Frequency compensation factor
 void windowFunction(QVector<double> &data, WindowingFunctionEnumT function, double &rCa, double &rCb)
 {
+    if(data.empty()) {
```
Comment (peterNordin) [inline HopsanGUI/Utilities/GUIUtilities.cpp:384]: Inst the problem here if data has size  1

File: HopsanGUI/Utilities/GUIUtilities.cpp (hunk 1 above)
Comment (peterNordin) [reply HopsanGUI/Utilities/GUIUtilities.cpp:421]: nevermind, I did not look closely enough, I though there were three out parameters rCa, rCb and rCc (but the last does note exist)

File: HopsanGenerator/src/generators/HopsanModelicaGenerator.cpp (hunk 3 above)
Comment (peterNordin) [reply HopsanGenerator/src/generators/HopsanModelicaGenerator.cpp]: OK

File: HopsanGUI/Utilities/GUIUtilities.cpp (hunk 4 above)
Comment (robbr48) [reply HopsanGUI/Utilities/GUIUtilities.cpp]: You are right. I changed the check so that data size must be greater than 1.

**Pull request merged**